File: pco_walrus_sync.rb
require 'base64'
require 'rest-client'
require 'json'
require 'date'
require 'midi-file'

# ================= CONFIG =================

PC_APP_ID = "YOUR_APP_ID"
PC_SECRET = "YOUR_SECRET"

PERSON_ID = "YOUR_PERSON_ID"
TARGET_SERVICE_TYPE_NAME = "Celebration Service"

BASE_URL = "https://api.planningcenteronline.com/services/v2"
MIDI_OUTPUT = "weekly_setlist_with_meta.mid"
MIDI_CHANNEL = 0

# ================= HELPERS =================

def auth_header
  token = Base64.strict_encode64("#{PC_APP_ID}:#{PC_SECRET}")
  { Authorization: "Basic #{token}" }
end

def get_json(url, params = {})
  response = RestClient.get(url, { params: params }.merge(auth_header))
  JSON.parse(response.body)
end

# ================= SERVICE TYPE =================

def get_service_type_id
  data = get_json("#{BASE_URL}/service_types")

  svc = data["data"].find do |s|
    s["attributes"]["name"].downcase == TARGET_SERVICE_TYPE_NAME.downcase
  end

  raise "Celebration Service not found." unless svc

  puts "Using service type: #{svc["attributes"]["name"]}"
  svc["id"]
end

# ================= PLAN FINDERS =================

def find_plan_by_date(service_type_id, target_date)
  url = "#{BASE_URL}/service_types/#{service_type_id}/plans"
  params = { order: "sort_date" }

  while url
    data = get_json(url, params)
    params = nil

    data["data"].each do |plan|
      sort_date = plan.dig("attributes", "sort_date")
      next unless sort_date

      plan_date = Date.parse(sort_date)

      puts "Checking plan: #{plan_date} - #{plan["attributes"]["title"]}"

      if plan_date.to_s == target_date
        puts "\nSelected plan: #{plan["attributes"]["title"]} - #{plan["attributes"]["dates"]}"
        return plan["id"]
      end
    end

    url = data.dig("links", "next")
  end

  nil
end

def find_next_scheduled_plan(service_type_id)
  today = Date.today
  data = get_json("#{BASE_URL}/service_types/#{service_type_id}/plans", { order: "sort_date" })

  data["data"].each do |plan|
    sort_date = plan.dig("attributes", "sort_date")
    next unless sort_date

    plan_date = Date.parse(sort_date)
    next if plan_date < today

    plan_id = plan["id"]

    team = get_json("#{BASE_URL}/service_types/#{service_type_id}/plans/#{plan_id}/team_members")

    team["data"].each do |member|
      person = member.dig("relationships", "person", "data", "id")
      if person == PERSON_ID
        puts "\nAuto-selected scheduled plan:"
        puts " #{plan["attributes"]["title"]} - #{plan["attributes"]["dates"]}"
        return plan_id
      end
    end
  end

  nil
end

# ================= SONG FETCH =================

def fetch_plan_songs_with_meta(service_type_id, plan_id)
  data = get_json(
    "#{BASE_URL}/service_types/#{service_type_id}/plans/#{plan_id}/items",
    { include: "arrangement,key,song" }
  )

  included = {}
  (data["included"] || []).each do |i|
    included["#{i["type"]}:#{i["id"]}"] = i
  end

  songs = []

  data["data"].each do |item|
    next unless item["attributes"]["item_type"].downcase == "song"

    title = item["attributes"]["title"]

    bpm = meter = key_name = nil

    arr_id = item.dig("relationships", "arrangement", "data", "id")
    if arr_id
      arr = included["arrangements:#{arr_id}"]
      bpm = arr.dig("attributes", "bpm")
      meter = arr.dig("attributes", "meter")
    end

    key_id = item.dig("relationships", "key", "data", "id")
    if key_id
      key = included["keys:#{key_id}"]
      key_name = key.dig("attributes", "name")
    end

    songs << { title: title, bpm: bpm, meter: meter, key: key_name }
  end

  songs
end

# ================= MIDI =================

def create_midi(songs)
  seq = MIDI::Sequence.new
  track = MIDI::Track.new(seq)
  seq.tracks << track

  track.events << MIDI::MetaEvent.new(MIDI::META_SEQ_NAME, "Celebration Service Setlist")
  track.events << MIDI::Tempo.new(MIDI::Tempo.bpm_to_mpq(120))

  puts "\nSongs in this set:\n"

  time = 0

  songs.each_with_index do |song, i|
    track.events << MIDI::ProgramChange.new(MIDI_CHANNEL, i, time)
    puts "#{i + 1}. #{song[:title]} | BPM:#{song[:bpm]} | TS:#{song[:meter]} | Key:#{song[:key]}"
    time += 480
  end

  File.open(MIDI_OUTPUT, 'wb') { |f| seq.write(f) }

  puts "\nMIDI file created: #{MIDI_OUTPUT}"
end

# ================= MAIN =================

puts "\n--- Celebration Service → Walrus Clock Sync (Ruby) ---\n"

svc_id = get_service_type_id

puts "\nChoose mode:"
puts "1 - Next plan I am scheduled for"
puts "2 - Pick a specific Sunday date"

print "\nEnter 1 or 2: "
choice = gets.strip

plan_id =
  if choice == "2"
    print "Enter Sunday date (YYYY-MM-DD): "
    date_input = gets.strip
    find_plan_by_date(svc_id, date_input)
  else
    find_next_scheduled_plan(svc_id)
  end

if plan_id.nil?
  puts "\nNo Celebration Service plan found."
  exit
end

songs = fetch_plan_songs_with_meta(svc_id, plan_id)

if songs.empty?
  puts "\nNo songs found in this plan."
  exit
end

create_midi(songs)

File: pco_walrus_sync.py
import base64
import requests
from datetime import datetime, timezone
from dateutil.relativedelta import relativedelta, SU
from midiutil import MIDIFile

# ================= CONFIG =================

PC_APP_ID = "ef6859b2d409c09b0de1df2788d4551ff10d01ae348e079b207cb58fc96ae984"
PC_SECRET = "pco_pat_c99952c219805b5ec88f2c72564e39931a310fcf7620f7f302015b05f35ea77ed2910211"

PERSON_ID = "AC63069036#"  # from Planning Center People profile URL
TARGET_SERVICE_TYPE_NAME = "Celebration Service"

BASE_URL = "https://api.planningcenteronline.com/services/v2"

MIDI_OUTPUT = "weekly_setlist_with_meta.mid"
MIDI_CHANNEL = 0

# ================= HELPERS =================

def auth_header():
    token = f"{PC_APP_ID}:{PC_SECRET}"
    token_b64 = base64.b64encode(token.encode()).decode()
    return {"Authorization": f"Basic {token_b64}"}

# ================= SERVICE TYPE =================

def get_service_type_id():
    headers = auth_header()
    resp = requests.get(f"{BASE_URL}/service_types", headers=headers)
    resp.raise_for_status()

    for svc in resp.json()["data"]:
        if svc["attributes"]["name"].lower() == TARGET_SERVICE_TYPE_NAME.lower():
            print("Using service type:", svc["attributes"]["name"])
            return svc["id"]

    raise Exception("Celebration Service not found.")

# ================= PLAN FINDERS =================

def find_plan_by_date(service_type_id, target_date):
    headers = auth_header()
    url = f"{BASE_URL}/service_types/{service_type_id}/plans"
    params = {"order": "sort_date"}

    while url:
        resp = requests.get(url, headers=headers, params=params)
        resp.raise_for_status()
        data = resp.json()

        plans = data.get("data", [])

        for plan in plans:
            sort_date = plan["attributes"]["sort_date"]
            if not sort_date:
                continue

            plan_date = datetime.fromisoformat(
                sort_date.replace("Z", "+00:00")
            ).date().isoformat()

            # DEBUG line (optional but helpful)
            print("Checking plan:", plan_date, "-", plan["attributes"]["title"])

            if plan_date == target_date:
                print("\nSelected plan:", plan["attributes"]["title"], "-", plan["attributes"]["dates"])
                return plan["id"]

        # move to next page if it exists
        url = data.get("links", {}).get("next")
        params = None  # only needed on first request

    return None



def find_next_scheduled_plan(service_type_id):
    headers = auth_header()

    params = {"order": "sort_date"}
    resp = requests.get(
        f"{BASE_URL}/service_types/{service_type_id}/plans",
        headers=headers,
        params=params
    )
    resp.raise_for_status()

    today = datetime.now(timezone.utc).date()

    for plan in resp.json()["data"]:
        sort_date = plan["attributes"]["sort_date"]
        if not sort_date:
            continue

        plan_date = datetime.fromisoformat(sort_date.replace("Z", "+00:00")).date()
        if plan_date < today:
            continue

        plan_id = plan["id"]

        team_resp = requests.get(
            f"{BASE_URL}/service_types/{service_type_id}/plans/{plan_id}/team_members",
            headers=headers
        )
        team_resp.raise_for_status()

        for member in team_resp.json()["data"]:
            rel = member["relationships"].get("person")
            if rel and rel["data"]["id"] == PERSON_ID:
                print("Auto-selected scheduled plan:")
                print(" ", plan["attributes"]["title"], "-", plan["attributes"]["dates"])
                return plan_id

    return None

# ================= SONG FETCH =================

def fetch_plan_songs_with_meta(service_type_id, plan_id):
    headers = auth_header()

    resp = requests.get(
        f"{BASE_URL}/service_types/{service_type_id}/plans/{plan_id}/items",
        headers=headers,
        params={"include": "arrangement,key,song"}
    )
    resp.raise_for_status()

    payload = resp.json()
    included = {f'{i["type"]}:{i["id"]}': i for i in payload.get("included", [])}

    songs_meta = []

    for item in payload["data"]:
        if item["attributes"]["item_type"].lower() != "song":
            continue

        title = item["attributes"]["title"]
        bpm = meter = key_name = None

        arr_rel = item["relationships"].get("arrangement")
        if arr_rel and arr_rel["data"]:
            arr = included.get(f'arrangements:{arr_rel["data"]["id"]}')
            if arr:
                bpm = arr["attributes"].get("bpm")
                meter = arr["attributes"].get("meter")

        key_rel = item["relationships"].get("key")
        if key_rel and key_rel["data"]:
            key = included.get(f'keys:{key_rel["data"]["id"]}')
            if key:
                key_name = key["attributes"].get("name")

        songs_meta.append({
            "title": title,
            "bpm": bpm,
            "meter": meter,
            "key": key_name
        })

    return songs_meta

# ================= MIDI =================

def create_midi(songs_meta):
    midi = MIDIFile(1)
    time = 0

    midi.addTrackName(0, 0, "Celebration Service Setlist")
    midi.addTempo(0, 0, 120)

    print("\nSongs in this set:")

    for i, song in enumerate(songs_meta):
        midi.addProgramChange(0, MIDI_CHANNEL, time, i)
        print(f'{i+1}. {song["title"]} | BPM:{song["bpm"]} | TS:{song["meter"]} | Key:{song["key"]}')
        time += 1

    with open(MIDI_OUTPUT, "wb") as f:
        midi.writeFile(f)

    print("\nMIDI file created:", MIDI_OUTPUT)

# ================= MAIN =================

if __name__ == "__main__":
    print("\n--- Celebration Service → Walrus Clock Sync ---\n")

    svc_id = get_service_type_id()

    print("\nChoose mode:")
    print("1 - Next plan I am scheduled for")
    print("2 - Pick a specific Sunday date")

    choice = input("\nEnter 1 or 2: ").strip()

    plan_id = None

    if choice == "2":
        date_input = input("Enter Sunday date (YYYY-MM-DD): ").strip()
        plan_id = find_plan_by_date(svc_id, date_input)
        if not plan_id:
            print("No Celebration Service plan found for that date.")
            exit()
    else:
        plan_id = find_next_scheduled_plan(svc_id)
        if not plan_id:
            print("No upcoming Celebration Service plans found where you are scheduled.")
            exit()

    songs_meta = fetch_plan_songs_with_meta(svc_id, plan_id)

    if not songs_meta:
        print("No songs found in this plan.")
        exit()

    create_midi(songs_meta)